src/mo_interpreter/interpret.ml

```diff
@@ -773,7 +773,8 @@ and define_id env id v =
 and define_pat env pat v =
   match pat.it with
   | WildP -> true
-  | LitP _ | SignP _ | AltP _ -> match_pat pat v <> None
+  | LitP _ | SignP _ -> match_pat pat v <> None
+  | AltP (pat1, pat2) -> define_pat env (if match_pat pat1 v <> None then pat1 else pat2) v
```

Comment:
```suggestion
  | AltP (pat1, pat2) -> define_pat env (if match_pat pat1 v <> None then pat1 else pat2) v
```
This recursive call to `match_pat` looks fishy to me. I think we maybe need to rethink why we do the separate `define_pat` pass in the first place. I wonder if it predates the definedness check that Joachim introduced and was used to define identifiers after initially black-holing them, to detect undefinedness dynamically. Maybe there is a better way.

Reply:
Indeed, looking at the code I think we could elimate define_pat using match_pat and a simpler map over the resulting environment to resolve all the promised recursive bindings (maybe). 
This would be both in the ir and source interpreters. Best done as a separate PR first, I think.

Reply:
https://github.com/dfinity/motoko/pull/4011 attempts to simplify this.

Reply:
I think once you mere 4011, there should be not change to the interpreter code.

Reply:
Yes, I noticed. The interpreters are now almost unaffected.